src/Client/HedgedConnections.cpp

```diff
@@ -0,0 +1,509 @@
+                is_first = false;
+            }
+        }
+    }
+
+    return addresses;
+}
+
+void HedgedConnections::sendCancel()
+{
+    std::lock_guard lock(cancel_mutex);
+
+    if (!sent_query || cancelled)
+        throw Exception("Cannot cancel. Either no query sent or already cancelled.", ErrorCodes::LOGICAL_ERROR);
+
+    for (auto & offset_status : offset_states)
+        for (auto & replica : offset_status.replicas)
+            if (replica.connection)
+                replica.connection->sendCancel();
+
+    cancelled = true;
+}
+
+Packet HedgedConnections::drain()
+{
+    std::lock_guard lock(cancel_mutex);
+
+    if (!cancelled)
+        throw Exception("Cannot drain connections: cancel first.", ErrorCodes::LOGICAL_ERROR);
+
+    Packet res;
+    res.type = Protocol::Server::EndOfStream;
+
+    while (!epoll.empty())
+    {
+        Packet packet = receivePacketImpl();
+        switch (packet.type)
+        {
+            case Protocol::Server::PartUUIDs:
+            case Protocol::Server::Data:
+            case Protocol::Server::Progress:
+            case Protocol::Server::ProfileInfo:
+            case Protocol::Server::Totals:
+            case Protocol::Server::Extremes:
+            case Protocol::Server::EndOfStream:
+                break;
+
+            case Protocol::Server::Exception:
+            default:
+                /// If we receive an exception or an unknown packet, we save it.
+                res = std::move(packet);
+                break;
+        }
+    }
+
+    return res;
+}
+
+Packet HedgedConnections::receivePacket()
+{
+    std::lock_guard lock(cancel_mutex);
+    return receivePacketUnlocked({});
+}
+
+Packet HedgedConnections::receivePacketUnlocked(AsyncCallback async_callback)
+{
+    if (!sent_query)
+        throw Exception("Cannot receive packets: no query sent.", ErrorCodes::LOGICAL_ERROR);
+    if (!hasActiveConnections())
+        throw Exception("No more packets are available.", ErrorCodes::LOGICAL_ERROR);
+
+    if (epoll.empty())
+        throw Exception("No pending events in epoll.", ErrorCodes::LOGICAL_ERROR);
+
+    return receivePacketImpl(std::move(async_callback));
+}
+
+Packet HedgedConnections::receivePacketImpl(AsyncCallback async_callback)
+{
+    int event_fd;
+    Packet packet;
+    bool finish = false;
+    while (!finish)
+    {
+        event_fd = getReadyFileDescriptor(async_callback);
+
+        if (fd_to_replica_location.contains(event_fd))
+        {
+            ReplicaLocation location = fd_to_replica_location[event_fd];
+            packet = receivePacketFromReplica(location, async_callback);
+            finish = true;
+        }
+        else if (timeout_fd_to_replica_location.contains(event_fd))
+        {
+            ReplicaLocation location = timeout_fd_to_replica_location[event_fd];
+            processTimeoutEvent(location, offset_states[location.offset].replicas[location.index].active_timeouts[event_fd]);
+        }
```

Comment:
Also, there may happen rare situation when we do not call `receivePacket` for a long time. Then, we can get data for next package **and** alarmed timer at the same time. Even if timer was alarmed a long after data was received. it is not guaranteed that we cannot get timer descriptor from epoll first, which will cause false positive connection timeout.

 A correct solution is to read all available descriptors first, then process it individually for each replica, skipping timers if we got replica fd already (this logic is used in `RemoteQueryExecutorReadContext::checkTimeoutImpl`).